A51-xds-custom-backend-metrics.md

```diff
@@ -0,0 +1,382 @@
+
+// Allow configuring minimum report interval. If not configured or badly configured non-positive, 
+// the default is 30s.
+// In the future, it might be useful to let users specify minimum report interval and default 
+// report interval separately. If the default report interval is not configured or badly configured 
+// non-positive, the default is 1 min. 
+// public OpenRcaServiceImpl(long minInterval, TimeUnit timeUnit, long 
+// defaultInterval, TimeUnit)
+public OrcaServiceImpl(long minReportInterval, TimeUnit unit);
+
+// Update the metrics value corresponding to the specified key.
+void setUtilizationMetric(String key, double value);
+
+// Replace the whole metrics data using the specified map.
+void setAllUtilizationMetrics(Map<String, Double> metrics);
+
+// Remove the metrics data entry corresponding to the specified key.
+void deleteUtilizationMetric(String key);
+
+// Update the CPU utilization metrics data.
+void setCPUUtilizationMetric(double value);
+
+// Clear the CPU utilization metrics data.
+void deleteCPUUtilizationMetric();
+
+// Update the memory utilization metrics data.
+void setMemoryUtilizationMetric(double value);
+
+// Clear the memory utilization metrics data.
+void deleteMemoryUtilizationMetric();
+}
+```
+
+All the methods above should be thread safe. 
+Allowing updating each metric key individually is convenient for a server application to adopt where it 
+comes with multiple isolated components and each generates their own set of metrics.
+
+A minimum reporting interval (30s by default) could be configured when creating the service.
+The minimum report interval is the lower bound of the OOB metrics report period. It means, in the 
+client OOB request message if the `report_interval` 
+goes below the service configured minimum interval then it is treated as the service minimum interval. 
+If `report_interval` is not specified (equal to 0), it is also treated as service minimum.
+There is no upper bound, the largest value you can specify is max integer.
+However, it might be useful to have a separate option called default report interval (default 1 minute) 
+that will be used when `report_interval` is not specified (equal to 0), see the code comment above.
+
+Alternatively we can use callback-style APIs in a pull model, but this is suboptimal because 
+it is not performance/scalability friendly.
+In a pull model, the time complexity of doing utilization measurement syscalls reaches O(M*N),
+M being the number of OOB clients and N being their frequency configuration.
+In a push model, the number of syscalls decouples with the number of OOB clients and their configurations.
+In fact, the server application is able to tune the performance by sampling the utilization data at a constant rate.
+
+Internally, the load report server will save each incoming client connection(`responseObserver`) 
+in a `RcaConnection` object or alike, with a custom frequency timer.
+
+```java
```

Comment:
I don't think we need Java-specific details here.  The language-agnostic description is sufficient.